equal_exprt now checks types of the operands

kroening opened the pull request:
This commit adds a precondition to the constructor of `equal_exprt`, which
enforces that the two operands have the same type.

A legacy method is offered to simplify the transition to the new behavior.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
equal_exprt now checks types of the operands

This commit adds a precondition to the constructor of equal_exprt, which
enforces that the two operands have the same type.

tautschnig commented on src/analyses/variable-sensitivity/constant_abstract_value.cpp:
```diff
@@ -146,7 +146,8 @@ abstract_value_pointert constant_abstract_valuet::constrain(
 exprt constant_abstract_valuet::to_predicate_internal(const exprt &name) const
 {
-  return equal_exprt(name, value);
+  // Temporary work around as name.type() need not be value.type().
+  return equal_exprt::unchecked(name, value);
```
That's probably a bug?

tautschnig replied:
I have fixed the bugs (well, at least those that the test suite surfaced).

tautschnig commented on src/util/std_expr.h:
```diff
@@ -1308,6 +1308,23 @@ class equal_exprt:public binary_relation_exprt
+    PRECONDITION(lhs().type() == rhs().type());
+  }
+
+  // Temporary workaround for callers that do not guarantee
+  // lhs.type() == rhs.type()
+  // This constructor will be removed.
+  equal_exprt(exprt _lhs, exprt _rhs, int)
+    : binary_relation_exprt(std::move(_lhs), ID_equal, std::move(_rhs))
+  {
+  }
+
+  // Temporary workaround for callers that do not guarantee
+  // lhs.type() == rhs.type()
+  // This function will be removed.
+  static equal_exprt unchecked(exprt _lhs, exprt _rhs)
+  {
+    return equal_exprt(std::move(_lhs), std::move(_rhs), 0);
   }
```
I'd strongly suggest we fix this right away rather than introducing an API that then needs to be removed via deprecation. If there is absolutely no way of debugging this now (I'm not sure why), then we need some mechanism of tracking this.

kroening replied:
Tried, and gave up.

kroening replied:
Tracking is easy: grep for `unchecked` will do it.

tautschnig replied:
Fix the fixes that I added this isn't necessary anymore. I'll leave it to you to decide whether you want to keep these (now unused) methods.

kroening replied:
ok, dropped

tautschnig commented on src/statement-list/parser.y:
```diff
@@ -1098,8 +1098,8 @@ Param_Assignment:
-      parser_stack($$) = equal_exprt{std::move(parser_stack($1)), 
-        std::move(parser_stack($3))};
+      parser_stack($$) = equal_exprt::unchecked(std::move(parser_stack($1)),
+        std::move(parser_stack($3)));
```
I'd say the statement-list code should just not be using `equal_exprt` for assignments, and should instead use `code_frontend_assignt`.

tautschnig replied:
I have added a commit to address this.

tautschnig committed:
Statement-list front-end: do not use equal_exprt for assignments

We have `code_frontend_assignt` for this purpose.

tautschnig committed:
Fix type inconsistencies in variable sensitivity

Abstract struct predicates have members with component type, not struct
type. Unit tests: concrete arrays over signedbv(32) have a subtype
that's signedbv(32).
